### ISSUE TITLE
CloseNotify called after ServeHTTP finished

### ISSUE THREAD

**Issue by admiral-Guck** [ipfs/go-ipfs#2521]
I am encountering a panic when running the daemon and doing certain ipfs operations:

````
Initializing daemon...
Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip4/192.168.1.114/tcp/4001
Swarm listening on /ip6/::1/tcp/4001
API server listening on /ip4/127.0.0.1/tcp/5001
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/8080
Daemon is ready
panic: net/http: CloseNotify called after ServeHTTP finished

goroutine 787 [running]:
panic(0x7f86e0, 0xc821589060)
	/usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:464 +0x3e6
net/http.(*response).CloseNotify(0xc8217e5040, 0x0)
	/usr/local/Cellar/go/1.6/libexec/src/net/http/server.go:1535 +0x9d
github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/prometheus/client_golang/prometheus.(*fancyResponseWriterDelegator).CloseNotify(0xc820092258, 0x7f2d60)
	/private/tmp/ipfs20160401-20845-k6fjou/src/github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/prometheus/client_golang/prometheus/http.go:219 +0x68
github.com/ipfs/go-ipfs/commands/http.internalHandler.ServeHTTP.func2(0x1507428, 0xc820092258, 0x1540258, 0xc82188d900, 0xc821588d30)
	/private/tmp/ipfs20160401-20845-k6fjou/src/github.com/ipfs/go-ipfs/commands/http/handler.go:136 +0x39
created by github.com/ipfs/go-ipfs/commands/http.internalHandler.ServeHTTP
	/private/tmp/ipfs20160401-20845-k6fjou/src/github.com/ipfs/go-ipfs/commands/http/handler.go:140 +0x49a
````

Using curl to connect to the webui yields this (any url will result in similar weirdness):
````
||
!˝%gR§¡ì‚®S>ß´¶0Ç"0
	*ÜHÜ˜
Ç0Ç
Çπ∂g„¢i5ÄíR	Tc-:Ûˇ2_XqÀq6wau˙è•9üS™∆∞0^ÿ ™V˙u\Ó,EG˙çÏMÙ◊˚∂!çî≤+Á«r[¿#r¸õ}€qòE<∏6ù∑€Â-4eÖÁÒxƒP∆=*T”œ‹~?#ê3H
U€›É#p‹'1a7⁄fﬁ;æ¶_i.1ZÀ¯TÛüîÆA£y-ˆÂüy˘6€÷m˙j´√Ó∆G∞◊“7d3m.∑V;ómÍëé;”—óA+∑rlÂá{:Äœ%©\∆ˇﬁ)∑C◊ï{\es˝ƒˆød“‡°Ûla#µ¬C§©Æ?®˜‘IÒP-256,P-384,P-521"AES-256,AES-128,Blowfish*
SHA256,SHA512
````

I can `ipfs swarm connect /ipfs/Qm...`, it exits successfully. The daemon crashes immediately after that, though.

At one time, running `ipfs swarm peers` printed these two addresses, then the daemon crashed.

````
/ip4/104.236.151.122/tcp/4001/ipfs/QmSoLju6m7xTh3DuokvT3886QRYqxAzb1kShaanJgW36yx
/ip4/178.62.158.247/tcp/4001/ipfs/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd
````

Other times, it lists all peers, exits sucessfully, and the daemon keeps running.

The Gateway Redirect plugin for Firefox always crashes the daemon before connecting to the gateway.
When the plugin is enabled, the daemon takes 10s to be ready, then crashes a couple seconds later. Otherwise it's done in 1s and answers certain requests w/o complaining.

**TLDR**: Daemon crashes rather eratically mostly when trying to do stuff over the network.
go 1.6, ipfs 0.3.11 with homebrew on 10.10.5.

May be related to #2108.
I'd gladly follow instructions to provide more information.

**Comment by area** [comment]
I'm seeing similar behaviour with go 1.5.3, ipfs 0.3.11 from homebrew on 10.11.3

**Comment by admiral-Guck** [comment]
Should have noted everything was running smoothly not long ago. Can't recall exactly when this started, though it was around the time I first installed the firefox plugin.

**Comment by whyrusleeping** [comment]
The close notify issues have been fixed in master. We will be shipping a new version soon which shouldn't have the mentioned bug. If you could try out latest master for me and report back, that would be super helpful

**Comment by admiral-Guck** [comment]
I've just tried 0.4.0-dev, all good!

Thanks

Jeromy Johnson:

**Comment by whyrusleeping** [comment]
This has been resolved
